resources/2025-11-11-secrets-redactdata.md

```diff
@@ -0,0 +1,756 @@
+
+**Reasons for rejection**:
+- **Cannot support recipes**: Recipe execution happens in backend async operation and needs access to sensitive data, this would break the recipe-based deployment.
+
+#### Option 2: Application-Layer Encryption with Backend Pre-Recipe Redaction  (Recommended)
+
+**Description**: The frontend encrypts sensitive fields marked with `x-radius-sensitive` using application-layer encryption before database storage. The backend controller loads the encrypted resource, decrypts sensitive fields into memory, immediately redacts the persisted record (saving `data: null` back to the database), and then runs the recipe using only the in-memory decrypted copy. The decrypted payload never leaves process memory. On failure, the controller returns `Result{Requeue=false}` to prevent retries.
+
+**Advantages**:
+- **Shortest exposure window**: Encrypted data exists in database momentarily (frontend save → backend redaction)
+- **Recipe compatible**: Recipe receives decrypted data via in-memory structures
+- **Type-agnostic**: Relies on `x-radius-sensitive` annotation, not specific resource types
+- **Multi-backend support**: Works with any recipe backend (Azure KV, HashiCorp, K8s)
+- **Application-layer protection**: Even if database encryption at rest is not configured, sensitive data is encrypted
+- **Testable**: Clear boundaries for testing encryption, decryption, redaction, and no-retry flow
+
+**Disadvantages**:
+- No automatic retries: failures require manual resubmission with fresh sensitive data
+- Application-layer encryption adds complexity: key management, encryption/decryption overhead, error handling
+- Performance overhead: Encryption/decryption adds latency to frontend save and backend processing
+- Requires updates across frontend, backend, controllers
+
+**Redaction timing and retries:**
+- **Frontend**: Encrypts fields marked with `x-radius-sensitive: true` before saving to database
+- **Backend**: Decrypts sensitive fields in memory, immediately persists sanitized copy (data: null), keeps decrypted data only in memory for recipe execution
+- **No retries on failure**: Controller returns `Result{Requeue:false}` for failure paths only, preventing retry attempts that would fail due to missing sensitive data
+- **Success path**: On successful completion, returns normal success result
+
+##### Application-Layer Encryption Details
+- Store root key in a Kubernetes secret (similar to ucp-cert)
+- Use Go's `crypto/cipher` package with ChaCha20-Poly1305 AEAD cipher
+- Generate unique nonce per encryption operation
+- Key rotation:
+    - **Initial implementation (Manual)**: 
+      - Administrator generates new 256-bit key
+      - Updates Kubernetes Secret `radius-encryption-key`
+      - Restarts Radius pods to reload key from secret
+      - Limitation: Cannot decrypt data encrypted with old key (acceptable since data lifetime is minimal (few seconds) (frontend save → backend redaction))
```

Comment:
> Cannot decrypt data encrypted with old key (acceptable since data lifetime is minimal (few seconds)

What is the user experience if decryption fails due to rotated key?